security/application.md

```diff
@@ -0,0 +1,412 @@
+run secure hashing functions quickly. Too quickly.
+
+The solution is to use a key derivation algorithm. These are much like normal
+hashing algorithms (they're actually quite different, but that difference is
+negligible), except they are intentionally slow.
+
+The three most common password hashing algorithms are bcrypt, scrypt, and
+PBKDF2. Each of these require a salt, but handle it themselves: the output of
+these functions is a string that contains the salt plus the hashed value. Store
+that entire string as the hashed password.
+
+Never compare hashed passwords using the built-in string equality function;
+this sets you up for a timing attack.
+
+### Timing attacks
+
+An attacker can learn a lot from _how long_ it takes to be denied access. If
+it's instant, that means the input didn't even pass validation; if it's kinda
+long, that means that the input got past validation and computed a hash but one
+of the first few characters of the hash were incorrect; a longer delay means
+that most of the hash was right. Knowing how much of the hash was right allows
+the attacker to narrow the attack space.
+
+The solution: use a constant-time equality check for comparing the hashed
+values. Bcrypt libraries ship with a function that does everything for you.
+ActiveSupport ships with `secure_compare` for constant-time comparisons. Worst
+case: pad the string to a fixed length then make sure your loop goes through
+every character even after you know the answer.
+
+## Encryption
+
+An encryption algorithm is one where a string can be made illegible and then
+returned back to the original string again, and where decrypting requires an
+out-of-band secret.
+
+Less abstractly: a string can be encrypted, and then to decrypt you must know
+the password.
+
+There are two kinds of encryption algorithms: symmetric and asymmetric.
+A symmetric algorithm is one where the same secret is used to encrypt it and
+decrypt it. An asymmetric algorithm is one where the string can be encrypted
+and decrypted using different secrets -- where the person encrypting cannot
+necessarily decrypt it.
+
+The most popular symmetric algorithms you'll encounter are AES and Twofish.
+These might be useful for encrypting a file to share with a group of people or
+for encrypting your filesystem. Rumor has it that 1Password uses something like
+it AES to encrypt an entire vault; it is encrypted at rest, and only decrypted
+when you enter the passphrase.
+
+Asymmetric encryption algorithms, also known as public/private keypair
+encryption, is more well-known -- in large part for how tricky they are to get
+right. Some famous ones are SSH, TLS (previously SSL), and PGP.  These start by
+generated a pair of encryption secrets known as the public and private keys.
+Anyone with the public key can encrypt a string, but only the holder of the
+private key can decrypt.
+
+(The math here is cool. I won't go into it.)
```

Comment:
If we're going to tease people with it, maybe a link to where they can read more?